### PR TITLE
py/mpprint.h: Add do_printf() macro.

### PR DESCRIPTION
**Pull request by IhorNehrutsa**
Move `do_printf()` from `drivers/esp-hosted/esp_hosted_hal.h` to `py/mpprint.h`

### CODE REVIEW

File: ports/renesas-ra/mpbthciport.c [hunk 1]
```diff
@@ -37,9 +39,6 @@
 
 #if MICROPY_PY_BLUETOOTH
 
-#define debug_printf(...) // mp_printf(&mp_plat_print, "mpbthciport.c: " __VA_ARGS__)
-#define error_printf(...) mp_printf(&mp_plat_print, "mpbthciport.c: " __VA_ARGS__)
```
Comment (projectgus) [inline ports/renesas-ra/mpbthciport.c:41]: These debug macros have lost the context of what file the log line is coming from.

**Comment by projectgus** [comment]
@IhorNehrutsa Could you please explain the reasoning behind this feature?

My understanding is that the API in `esp_hosted_hal.h` is for integrating an arbitrary program with the ESP Hosted logging functions, so it's designed around ESP Hosted (and ESP-IDF ) logging conventions. Moving that ESP Hosted feature (and design) to become a core part of MicroPython is a big change.

Suggest a good place to start is these two [questions from last time](https://github.com/micropython/micropython/pull/8050#issuecomment-1734746745):

* What problem does this solve?
* Can you please walk step by step through a hypothetical example that explains how a developer would use this to find and debug some MicroPython behaviour?

Plus two new questions:

* Why choose this API for MicroPython logging?
* What's the plan for documenting this API and ensuring it's adopted, if it's going to be a core part of MP?

File: drivers/esp-hosted/esp_hosted_hal.h [hunk 2]
```diff
@@ -29,36 +29,6 @@
 #ifndef MICROPY_INCLUDED_DRIVERS_ESP_HOSTED_HAL_H
 #define MICROPY_INCLUDED_DRIVERS_ESP_HOSTED_HAL_H
 
-#ifndef ESP_HOSTED_DEBUG
-#define ESP_HOSTED_DEBUG    (0)
-#endif
-
-#if ESP_HOSTED_DEBUG
-#define PROTOBUF_C_UNPACK_ERROR(...)    error_printf(__VA_ARGS__);
-#endif
```
Comment (iabdalkader) [inline drivers/esp-hosted/esp_hosted_hal.h:38]: This block shouldn't be removed.

**Comment by iabdalkader** [comment]
> My understanding is that the API in `esp_hosted_hal.h` is for integrating an arbitrary program with the ESP Hosted logging functions, so it's designed around ESP Hosted (and ESP-IDF ) logging conventions

Those macros were actually just meant to be used internally by the esp-hosted driver as a debugging aid, they are also unrelated to ESP-IDF. It was just becoming hard to see useful error output from debugging, warnings etc.. and I wanted some colorful output :) It would be nice if they are standardized in MicroPython, and replace all internal debugging macros/helpers at some point, but they could probably use some work first. Re the context, it can be maintained by using `__FILE__, `__LINE__` etc..

File: drivers/esp-hosted/esp_hosted_hal.h (hunk 2 above)
Comment (IhorNehrutsa) [reply drivers/esp-hosted/esp_hosted_hal.h]: Reverted

File: ports/renesas-ra/mpbthciport.c (hunk 1 above)
Comment (IhorNehrutsa) [reply ports/renesas-ra/mpbthciport.c:41]: See  do_print_line()

**Comment by dpgeorge** [comment]
Thanks for the contribution, but we don't need all this additional complexity in MicroPython.  I know it's not adding any code size, but it adds lines to the source files and that adds to the complexity for someone reading and trying to understand the code.

The printf-style debugging that we do have in MicroPython is already plenty, and should be done using local-to-the-file `DEBUG_printf()` macros.

**Pull request closed without merging**